framework/src/samplers/Sampler.C

```diff
@@ -12,103 +12,79 @@
+  params += validParams<DistributionInterface>();
+
+  params.addClassDescription("A base class for distribution sampling.");
   params.addRequiredParam<std::vector<DistributionName>>(
-      "distributions",
-      "The names of distributions that you want to use to perturb the given parameters");
+      "distributions", "The names of distributions that you want to sample.");
+  params.addParam<unsigned int>("seed", 0, "Random number generator initial seed");
   params.registerBase("Sampler");
   return params;
 }
 
 Sampler::Sampler(const InputParameters & parameters)
   : MooseObject(parameters),
-    RandomInterface(parameters,
-                    *parameters.get<FEProblemBase *>("_fe_problem_base"),
-                    parameters.get<THREAD_ID>("_tid"),
-                    false),
+    SetupInterface(this),
     DistributionInterface(this),
-    Restartable(parameters, "Samplers"),
-    _tid(getParam<THREAD_ID>("_tid")),
-    _reseed_for_new_sample(getParam<bool>("reseed")),
-    _dist_names(getParam<std::vector<DistributionName>>("distributions")),
-    _var_names(getParam<std::vector<std::string>>("perturb_parameters")),
-    _current_sample(0),
-    _failed_runs(true)
+    _distribution_names(getParam<std::vector<DistributionName>>("distributions")),
+    _seed(getParam<unsigned int>("seed"))
 {
-  _var_dist_map.clear();
-  _probability_weight.clear();
-  _var_value_hist.clear();
-  if (_var_names.size() != _dist_names.size())
-    mooseError("The size of perturb_parameters (",
-               _var_names.size(),
-               ") != the size of distributions (",
-               _dist_names.size(),
-               ").");
-  if (!_var_names.empty())
-  {
-    for (unsigned int i = 0; i < _var_names.size(); ++i)
-    {
-      _var_dist_map[_var_names[i]] = &getDistributionByName(_dist_names[i]);
-      _var_value_map[_var_names[i]] = 0.0;
-      _var_value_hist[_var_names[i]] = std::vector<Real>();
-    }
-  }
+  _generator.seed(0, _seed);
+  for (const DistributionName & name : _distribution_names)
+    _distributions.push_back(&getDistributionByName(name));
+  setNumberOfRequiedRandomSeeds(1);
 }
 
 void
-Sampler::generateSamples()
+Sampler::execute()
 {
+  // Get the samples then save the state so that subsequent calls to getSamples returns the same
+  // random numbers until this execute command is called again.
+  getSamples();
+  _generator.saveState();
 }
 
-std::vector<std::string>
-Sampler::getSampledVariableNames()
+std::vector<DenseMatrix<Real>>
+Sampler::getSamples()
 {
-  return _var_names;
-}
-
-std::vector<Real>
-Sampler::getSampledValues(const std::vector<std::string> & variableNames)
-{
-  std::vector<Real> vals;
-  for (auto & var_name : variableNames)
-  {
-    Real val = getSampledValue(var_name);
-    vals.push_back(val);
-  }
-  return vals;
+  _generator.restoreState();
+  sampleSetUp();
+  std::vector<DenseMatrix<Real>> output(_distributions.size());
+  for (std::size_t i = 0; i < _distributions.size(); ++i)
```

Comment:
It's important to understand that `beginIndex()` is not an external function. It's a macro. Likewise, `auto` is determined at compile time so there are no speed penalties for using this syntax.

I originally hated `auto` too but it removes the ability to even create errors. It turns C++ into Python where the type **always** comes from the RHS:
```
  v = []
  d = {}
  a = b
  # It's not even possible to declare a type on the LHS, thus you've never made a type error with Python.
```

As far as readability, I argue against that as well. Naming fixes most of that. "i, j, k" are always loop indexes in any sane code. You still have to use `const`, `&`, and `*` when using `auto` so you still know a lot about type you are working with. The best part is, if you use `auto` everywhere you never have to go back and fixup types if you change the declaration. We do this every day with Python.